quinn-proto/src/connection.rs

```diff
@@ -2595,37 +2590,49 @@ where
 
         // STREAM
         while buf.len() + frame::Stream::SIZE_BOUND < max_size && space_id == SpaceId::Data {
-            let mut stream = match self.streams.pending.pop_front() {
+            let max_data_len = max_size as usize - buf.len() - frame::Stream::SIZE_BOUND;
+            let id = match self.streams.pending.pop() {
                 Some(x) => x,
                 None => break,
             };
-            if self
-                .streams
-                .send_mut(stream.id)
-                .map_or(true, |s| s.state.was_reset())
-            {
-                self.unacked_data -= stream.data.len() as u64;
-                continue;
-            }
-            let len = cmp::min(
-                stream.data.len(),
-                max_size as usize - buf.len() - frame::Stream::SIZE_BOUND,
-            );
-            let data = stream.data.split_to(len);
-            let fin = stream.fin && stream.data.is_empty();
-            trace!(id = %stream.id, off = stream.offset, len, fin, "STREAM");
-            let frame = frame::Stream {
-                id: stream.id,
-                offset: stream.offset,
-                fin,
-                data,
+            let stream = match self.streams.send_mut(id) {
+                Some(s) => s,
+                // Stream was reset with pending data
+                None => continue,
             };
-            frame.encode(true, buf);
-            if !stream.data.is_empty() {
-                stream.offset += len as u64;
-                self.streams.pending.push_front(stream);
-            }
-            stream_frames.push(frame);
+            match stream.state {
```

Comment:
This seems like a big transformation that I can't fully follow. Where does all this code come from?

Reply:
Rearranged the send window refactoring and simplified the state checks to make this particular change a bit smaller. Should hopefully be more obvious now; the remaining changes are direct consequences of the move from using prebaked stream frames to composing them based on buffer and finish state on the fly.

Reply:
Yeah, I think it's better now. Let's merge this thing already!